pkg/storage/oss/oss.go

```diff
@@ -0,0 +1,660 @@
+		return nil, err
+	}
+
+	return map[string]string{
+		imageRegistrySecretDataKey: string(confData),
+	}, nil
+}
+
+func (d *driver) sharedCredentialsDataFromStaticCreds() ([]byte, error) {
+	if d.credentials == nil || d.credentials.accessKeyID == "" || d.credentials.accessKeySecret == "" {
+		return []byte{}, fmt.Errorf("invalid credentials for Alibaba Cloud")
+	}
+	buf := &bytes.Buffer{}
+	fmt.Fprint(buf, "[default]\n")
+	fmt.Fprintf(buf, "alibabacloud_access_key_id = %s\n", d.credentials.accessKeyID)
+	fmt.Fprintf(buf, "alibabacloud_secret_access_key = %s\n", d.credentials.accessKeySecret)
+
+	return buf.Bytes(), nil
+}
+
+// bucketExists checks whether or not the OSS bucket exists
+func (d *driver) bucketExists(bucketName string) (bool, error) {
+	if len(bucketName) == 0 {
+		return false, nil
+	}
+
+	svc, err := d.getOSSService()
+	if err != nil {
+		return false, err
+	}
+
+	_, err = svc.GetBucketInfo(bucketName)
+
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// StorageExists checks if an OSS bucket with the given name exists
+// and we can access it
+func (d *driver) StorageExists(cr *imageregistryv1.Config) (bool, error) {
+	if len(d.Config.Bucket) == 0 {
+		return false, nil
+	}
+
+	bucketExists, err := d.bucketExists(d.Config.Bucket)
+	if bucketExists {
+		util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionTrue, "OSS Bucket Exists", "")
+	}
+	if err != nil {
+		if oerr, ok := err.(oss.ServiceError); ok {
+			util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionFalse, oerr.Code, oerr.Error())
+			return false, nil
+		}
+		util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionUnknown, "Unknown Error Occurred", err.Error())
+		return false, err
+	}
+
+	return true, nil
+
+}
+
+// StorageChanged checks to see if the name of the storage medium
+// has changed
+func (d *driver) StorageChanged(cr *imageregistryv1.Config) bool {
+	if !reflect.DeepEqual(cr.Status.Storage.OSS, cr.Spec.Storage.OSS) {
+		util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionUnknown, "OSS Configuration Changed", "OSS storage is in an unknown state")
+		return true
+	}
+
+	return false
+}
+
+// CreateStorage attempts to create an OSS bucket
+// and apply any provided tags
+func (d *driver) CreateStorage(cr *imageregistryv1.Config) error {
+	infra, err := util.GetInfrastructure(d.Listers)
+	if err != nil {
+		return err
+	}
+
+	if err := d.UpdateEffectiveConfig(); err != nil {
+		return err
+	}
+
+	svc, err := d.getOSSService()
+	if err != nil {
+		return err
+	}
+
+	// If a bucket name is supplied, and it already exists and we can access it
+	// just update the config
+	var bucketExists bool
+	if len(d.Config.Bucket) != 0 {
+		bucketExists, err = d.bucketExists(d.Config.Bucket)
+		if bucketExists {
+			util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionTrue, "OSS Bucket Exists", "")
+		}
+		if err != nil {
+			if oerr, ok := err.(oss.ServiceError); ok {
```

Comment:
@menglingwei I tested this today. When the configuration is handled for the first time the bucket will not exist. The expectation is that the bucket will be created. In the current implementation the `err` is returned with the message of `NoSuchBucket`.

I updated this code to test the provider. My code is below:
```
	var bucketExists bool
	if len(d.Config.Bucket) != 0 {
		bucketExists, err = d.bucketExists(d.Config.Bucket)
		if bucketExists {
			util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionTrue, "OSS Bucket Exists", "")
		}
		if oerr, ok := err.(oss.ServiceError); ok {
			switch oerr.Code {
			case "NoSuchBucket", "NotFound":
				klog.Infof("The bucket %s not found ", d.Config.Bucket)
				// util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionFalse, oerr.Code, oerr.Error())
				// return fmt.Errorf("The bucket %s not found ", d.Config.Bucket)
			default:
				util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionUnknown, "Unknown Error Occurred", err.Error())
				return err
			}
		} else if err != nil {
			util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionUnknown, "Unknown Error Occurred", err.Error())
			return err
		}
	}
```

Please update this logic so that the bucket gets created when it is not found. Thanks!

Reply:
I updated the logic. If the error code is "NoSuchBucket" or "NotFound", just write log.